bin/install

```diff
@@ -1,105 +1,149 @@
 #!/usr/bin/env bash
 
+# We have no need for filename expansion / globbing,
+# so we'll disable it here.
+# We'll take advantage of the safety this confers when doing something like:
+#
+#   confg_args=(--some arg $USER_EXTRA_ARGS)
+#
+# Without set -f, $USER_EXTRA_ARGS can be subject to any glob patterns therein!
+set -f
+
+# fail-fast
+set -euo pipefail
+
 install_luaJIT() {
   local install_type=$1
   local raw_version=$2
   local install_path=$3
-  local luaJIT_version=$(echo $raw_version | sed -e 's/--.*$//g')
-  local lua_rocks_version=$(echo $raw_version | sed -e "s/^.*--//g")
+  local luaJIT_version="${raw_version%%--*}"
+  local lua_rocks_version="${raw_version##*--}"
 
-  if [ "$TMPDIR" = "" ]; then
-    local tmp_download_dir=$(mktemp -d)
+  local tmp_download_dir
+  if [[ "${TMPDIR-}" == "" ]]; then
+    tmp_download_dir=$(mktemp -d)
   else
-    local tmp_download_dir=$TMPDIR
+    tmp_download_dir=$TMPDIR
   fi
 
-  local luaJIT_source_path=$(get_luaJIT_download_file_path $install_type $luaJIT_version $tmp_download_dir)
-  local lua_rocks_source_path=$(get_lua_rocks_download_file_path $install_type $lua_rocks_version $tmp_download_dir)
-  download_luaJIT_source $install_type $luaJIT_version $luaJIT_source_path
-  download_lua_rocks_source $install_type $lua_rocks_version $lua_rocks_source_path
+  # canonicalize the path.
+  # this has several benefits:
+  #  1. if the luajit/luarocks build-systems also does this (hint: they do),
+  #     then the path we use here will be guaranteed to be identical.
+  #  2. this removes silly stuff like //, /./ throughout the path,
+  #     and removes a trailing slash (e.g. macOS's $TMPDIR).
+  tmp_download_dir=$(realpath -- "$tmp_download_dir")
+
+  local luaJIT_source_path
+  luaJIT_source_path=$(get_luaJIT_download_file_path "$install_type" "$luaJIT_version" "$tmp_download_dir")
```

Comment:
This is split into two lines, because while the `get_luaJIT_download_file_path "$install_type" "$luaJIT_version" "$tmp_download_dir` part may fail, the preceding `local` will succeed (sounds ridiculous, I know). Consequently, something like this will result in any failure within `get_luaJIT_download_file_path` being hidden:

```bash
local luaJIT_source_path=$(get_luaJIT_download_file_path "$install_type" "$luaJIT_version" "$tmp_download_dir")
````

but the following expression causes the script to exit (assuming `set -e`, as we have above) if such an error occurs:

```bash
local luaJIT_source_path
luaJIT_source_path=$(get_luaJIT_download_file_path "$install_type" "$luaJIT_version" "$tmp_download_dir")
````

Reply:
The shellcheck wiki entry on what/why: https://www.shellcheck.net/wiki/SC2155